src/main/java/org/sharedid/endpoint/handler/CheckOptOutHandler.java

```diff
@@ -0,0 +1,61 @@
+package org.sharedid.endpoint.handler;
+
+import io.vertx.core.Handler;
+import io.vertx.core.http.Cookie;
+import io.vertx.ext.web.RoutingContext;
+import org.sharedid.endpoint.context.DataContext;
+import org.sharedid.endpoint.util.ExtraHeadersCookie;
+import org.sharedid.endpoint.util.ResponseUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Component
+public class CheckOptOutHandler implements Handler<RoutingContext> {
+    private static final Logger logger = LoggerFactory.getLogger(CheckOptOutHandler.class);
+
+    private String optOutCookieValue;
+    private String sharedIdCookieName;
+    private Long sharedIdOptOutTtl;
+    private Boolean isSecureCookiesEnabled;
+
+    @Autowired
+    public CheckOptOutHandler(@Value("${cookie.shared-id.opt-out-value}") String optOutCookieValue,
+                              @Value("${cookie.shared-id.name}") String sharedIdCookieName,
+                              @Value("${cookie.shared-id.opt-out-ttl}") Long sharedIdOptOutTtl,
+                              @Value("${cookies.secure}") Boolean isSecureCookiesEnabled) {
+        this.optOutCookieValue = optOutCookieValue;
+        this.sharedIdCookieName = sharedIdCookieName;
+        this.sharedIdOptOutTtl = sharedIdOptOutTtl;
+        this.isSecureCookiesEnabled = isSecureCookiesEnabled;
+    }
+
+    @Override
+    public void handle(RoutingContext routingContext) {
+        DataContext dataContext = DataContext.from(routingContext);
+
+        String userId = dataContext.getUserId();
+
+        boolean hasOptedOutUserId = optOutCookieValue.equals(userId);
+
+        if (hasOptedOutUserId || dataContext.getHasOptOutCookie()) {
```

Comment:
Since `ProcessOptInHandler` doesn't change value of `dataContext.getHasOptOutCookie()` which is from cookie `"${cookie.opt-out.name}"`, which means if cookie `"${cookie.opt-out.name}"` is not blank for a user, then this user will keep in opt out status even the user tries to opt in.

Reply:
ProcessOptInHandler sets the cookie's age to 0, which causes the browser to delete the cookie. On the next request to shraed id, it will see that the user doesn't have a cookie, and generate a new id.

Reply:
@coryhammon1 my concern is not about cookie: `cookie.shared-id.name=sharedid`
my concern is about cookie: `cookie.opt-out.name=sharedid_optout`
I only find code read value from this cookie, but can't find any code set value for this cookie.
2 things need consider:
1) since there is no code set value for cookie: `cookie.opt-out.name=sharedid_optout`, which means for normal user, this cookie is always empty. So all the logic around that cookie can be removed;
2) since code read the cookie exists, if some user (shouldn't exist in real world) manually add that cookie in browser, then the user always stuck in opt out mode;

Reply:
Got it. The opt out cookie was legacy code that never got removed. The opt out is exclusively determined by the cookie value being zeroed out. I added a commit that removed the opt out cookie entirely.

Reply:
cool.